### PR TITLE
[lora, trainer] feat: Add lora support for VLM trainer

### PR DESCRIPTION
**Pull request by wyettzeng**
### What does this PR do?

Add lora support for VLM trainer

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> Validation results (training curves, eval metrics) for changes not covered by CI.

### API and Usage Example

> Show API changes and usage examples if applicable.

### Design & Code Changes

> High-level design description and specific change list.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

**Review by Coach257 — CHANGES_REQUESTED** [review]
**`freeze_vit` vs vision LoRA**

Current order (`_setup_lora()` then `visual.requires_grad_(False)`) freezes vision `lora_A`/`lora_B` whenever `freeze_vit=True`. That fights adapter training.

Suggested rule:
- If the visual tower already has LoRA injected → **ignore `freeze_vit` entirely** (leave `_setup_lora()` as-is: base frozen, vision adapters trainable).
- If visual has **no** LoRA → apply existing `freeze_vit` policy (full-FT: freeze backbone, keep merger/projector trainable on Omni; LLM-only LoRA: freeze the whole visual).

Please gate on “vision has LoRA”, not bare `lora_enabled`. Align FLOPs `vision_lora_enabled` with whether vision adapters are actually trainable, not `not freeze_vit`.

File: veomni/trainer/vlm_trainer.py [hunk 1]
```diff
@@ -174,11 +191,23 @@ def _freeze_model_module(self):
...
+                )
+                audio_proj.requires_grad_(True)
+
+        has_trainable_lora = any(
+            param.requires_grad and (".lora_A." in name or ".lora_B." in name)
+            for name, param in self.base.model.named_parameters()
+        )
+        if lora_enabled and not has_trainable_lora:
+            raise ValueError(
+                "VLM LoRA configuration produced no trainable adapters after applying "
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply veomni/trainer/vlm_trainer.py]: a bit out of the scope for this pr

File: veomni/trainer/callbacks/trace_callback.py [hunk 2]
```diff
@@ -189,6 +189,9 @@ def __init__(self, trainer: "BaseTrainer") -> None:
 
         args: "VeOmniArguments" = self.trainer.args
         self.lora_config = trainer.model.get_lora_config() if hasattr(trainer.model, "get_lora_config") else None
+        self.vision_lora_enabled = None
+        if self.lora_config is not None and hasattr(args.train, "freeze_vit"):
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply veomni/trainer/callbacks/trace_callback.py]: changed

File: configs/multimodal/qwen3_5_moe/qwen3_5_moe_vl_lora.yaml [hunk 3]
```diff
@@ -49,7 +55,7 @@ train:
       fsdp_mode: fsdp2
   gradient_checkpointing:
     enable_reentrant: false
-  freeze_vit: true
+  freeze_vit: false
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply configs/multimodal/qwen3_5_moe/qwen3_5_moe_vl_lora.yaml]: freeze_vit is no longer looked at when lora is enabled

File: veomni/trainer/vlm_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wyettzeng) [reply veomni/trainer/vlm_trainer.py]: you know what, you convinced me, changed

File: tests/lora/test_vlm_lora.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        False,
+        False,
+        id="qwen3_5_moe",
+    ),
+    pytest.param(
+        "configs/multimodal/qwen3_vl/qwen3_vl_moe_lora.yaml",
+        "tests/toy_config/qwen3vlmoe_toy/config.json",
+        True,
+        False,
+        id="qwen3_vl_moe",
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply tests/lora/test_vlm_lora.py:31]: tower freezing no longer looked at for lora-enabled config

File: tests/models/test_vlm_trainer.py [hunk 5]
```diff
@@ -24,22 +24,10 @@
...
 @pytest.mark.parametrize("config_path", _FREEZE_VIT_VLM_CASES)
 def test_freeze_vit_on_vlm_model(config_path, freeze_vit):
-    # This test only constructs the model on `meta` and verifies freeze
-    # behaviour — it never runs forward. Use an all-eager ops config so the
-    # build works everywhere: it pins every per-op field (including the
-    # Qwen3.5 GatedDeltaNet trio that has no FLA backend on NPU and the
-    # GPU-only liger/triton defaults that fail NPU validation). Eager paths
-    # that raise only at forward time are fine because this test never
-    # forwards.
+    # Structural meta-device coverage; eager ops keep it accelerator-agnostic.
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply tests/models/test_vlm_trainer.py]: done

File: veomni/utils/count_flops.py [hunk 6]
```diff
@@ -872,15 +889,20 @@ def _estimate_qwen_vit_flop(self, images_seqlens, config, lora_config=None):
...
-            module_shapes=vision_module_shapes,
-            detached_without_targets=True,
-        )
         target_modules = lora_config.target_modules if lora_config is not None else ()
-        vision_has_lora = any(module in vision_module_shapes for module in target_modules or ())
+        vision_has_lora = vision_lora_enabled is not False and any(
+            module in vision_module_shapes for module in target_modules or ()
+        )
+        if lora_config is not None and not vision_has_lora:
+            dense_N_flops = 2 * dense_N * tokens_sum
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply veomni/utils/count_flops.py:905]: conv3d needed for factor of 4 since we need to compute the gradient

**Review by Coach257 — APPROVED** [review]

**Comment by Coach257** [comment]
TODO: A more appropriate approach to lora fintune vlm is training vit_merger & llm_lora. Support it in omni_v2 through module-level lora setup & freeze_step.
#988
